packages/block-library/src/query/edit/query-block-setup.js

```diff
@@ -20,29 +23,35 @@ const QueryBlockSetup = ( {
 	name: blockName,
 } ) => {
 	const { postType, inherit } = query;
-	const { replaceInnerBlocks } = useDispatch( blockEditorStore );
+	const { replaceBlocks, replaceInnerBlocks } = useDispatch(
+		blockEditorStore
+	);
 	const { postTypesSelectOptions } = usePostTypes();
 	const updateQuery = ( newQuery ) =>
 		setAttributes( { query: { ...query, ...newQuery } } );
-	const onFinish = ( innerBlocks ) => {
-		if ( innerBlocks ) {
-			replaceInnerBlocks(
-				clientId,
-				createBlocksFromInnerBlocksTemplate( innerBlocks ),
-				false
-			);
-		}
-	};
 	const onVariationSelect = ( nextVariation ) => {
 		if ( nextVariation.attributes ) {
 			setAttributes( nextVariation.attributes );
 		}
 		if ( nextVariation.innerBlocks ) {
-			onFinish( nextVariation.innerBlocks );
+			replaceInnerBlocks(
+				clientId,
+				createBlocksFromInnerBlocksTemplate(
+					nextVariation.innerBlocks
+				),
+				false
+			);
 		}
 	};
 	const onBlockPatternSelect = ( blocks ) => {
-		onFinish( [ [ 'core/query-loop', {}, blocks ] ] );
+		const clonedBlocks = blocks.map( ( block ) => cloneBlock( block ) );
+		// We need to override the attributes that can be set in the Placeholder.
+		Object.assign( clonedBlocks[ 0 ].attributes.query, {
```

Comment:
do you think we should ensure that there's only one block and that it's a "query"?

Reply:
I have addressed this here: https://github.com/WordPress/gutenberg/pull/30471/commits/0c7683a3e2290b2734a647bd1d614c7316192f4b.

I now search in top level blocks though for Query and properly handle the needed changes by deep cloning these attributes.